### PR TITLE
Add xBit provider

### PR DESCRIPTION
**Pull request by duramato**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

Implements https://github.com/pymedusa/Medusa/issues/3095

### CODE REVIEW

File: medusa/providers/torrent/json/xbit.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+        # Cache
+        self.cache = tv.Cache(self, min_time=10)  # Only poll xBiT every 10 minutes max
+
+    def search(self, search_strings, age=0, ep_obj=None, **kwargs):
+        """
+        Search a provider and parse the results.
+
+        :param search_strings: A dict with mode (key) and the search value (value)
+        :param age: Not used
```
Comment (p0psicles) [inline medusa/providers/torrent/json/xbit.py]: You don't need to pass these parameters anymore.

File: medusa/providers/torrent/json/xbit.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                    'pubdate': pubdate,
+                }
+                if mode != 'RSS':
+                    log.debug('Found result: {0} with {1} seeders and {2} leechers',
+                              title, seeders, leechers)
+
+                items.append(item)
+            except (AttributeError, TypeError, KeyError, ValueError, IndexError):
+                log.error('Failed parsing provider. Traceback: {0!r}',
+                          traceback.format_exc())
```
Comment (p0psicles) [inline medusa/providers/torrent/json/xbit.py]: Use log.exception

**Comment by OmgImAlexis** [comment]
@duramato can this be rebased and any updates on when it can be merged?

**Comment by p0psicles** [comment]
Can you update change log?

**Comment by duramato** [comment]
This provider doesn't provider any favicon or icon, what should we do?

**Comment by p0psicles** [comment]
Something like this?
xbit

**Comment by duramato** [comment]
Seems good to me

File: medusa/providers/torrent/json/xbit.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        # URLs
+        self.url = 'https://xbit.pw'
+        self.urls = {
+            'search': urljoin(self.url, '/api.php')
+        }
+
+        # Proper Strings
+
+        # Miscellaneous Options
+        self.freeleech = False
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:38]: self.freeleech is not sued and can be removed

File: medusa/providers/torrent/json/xbit.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        # Torrent Stats
+        self.minseed = None
+        self.minleech = None
+
+        # Cache
+        self.cache = tv.Cache(self, min_time=10)  # Only poll xBiT every 10 minutes max
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:45]: `# Only poll xBiT every 10 minutes max` can be removed

File: medusa/providers/torrent/json/xbit.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            for separator in string_separator:
+                for search_string in search_strings[mode]:
+                    search_url = self.urls['search']
+                    if mode != 'RSS':
+                        log.debug('Search string: {search}',
+                                  {'search': search_string})
+                        # Replaces spaces with either a dot or a plus
+                        # Needed in order for it to return all results
+                        search_string = search_string.replace(' ', separator)
+                    search_params = {'search': search_string,
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:69]: Please init search_params outside the for loop (like the other providers)

File: medusa/providers/torrent/json/xbit.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+            for separator in string_separator:
+                for search_string in search_strings[mode]:
+                    search_url = self.urls['search']
+                    if mode != 'RSS':
+                        log.debug('Search string: {search}',
+                                  {'search': search_string})
+                        # Replaces spaces with either a dot or a plus
+                        # Needed in order for it to return all results
+                        search_string = search_string.replace(' ', separator)
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:68]: This should be done before the log.debug

File: medusa/providers/torrent/json/xbit.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        torrent_rows = data.get('dht_results')
+        if not torrent_rows:
+            log.debug('Data returned from provider does not contain any torrents')
+            return items
+        for row in torrent_rows[:-1]:
+
+            try:
+                title = row['NAME']
+                download_url = row['MAGNET']
+                download_url = download_url + self._custom_trackers
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:105]: this should be after the check

File: medusa/providers/torrent/json/xbit.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                if not all([title, download_url]):
+                    continue
+
+                seeders = 1  # Provider does not provide seeders
+                leechers = 0  # Provider does not provide leechers
+
+                # Filter unseeded torrent
+                if seeders < min(self.minseed, 1):
```
Comment (medariox) [inline medusa/providers/torrent/json/xbit.py:113]: outdated, see other providers

**Comment by medariox** [comment]
After testing this provider for a bit, I must say I'm against adding it. From my experience it times out very often, the search functionality is really limited and there are so many better alternatives. I see absolutely no value in adding this provider atm.

**Comment by duramato** [comment]
I'm neutral about it, it's not a provider I desperatly need. I made the PR at that time to implement the feature request.

**Pull request closed without merging**